pkg/report/sarif.go

```diff
@@ -0,0 +1,174 @@
+	"io"
+	"regexp"
+	"strings"
+
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/vulnerability"
+
+	"github.com/owenrumney/go-sarif/v2/sarif"
+)
+
+// regex to extract file path in case string includes (distro:version)
+var re = regexp.MustCompile(`(?P<path>.+?)(?:\s*\((?:.*?)\).*?)?$`)
+
+// SarifWriter implements result Writer
+type SarifWriter struct {
+	Output  io.Writer
+	Version string
+	run     *sarif.Run
+}
+
+type sarifData struct {
+	vulnerabilityId  string
+	fullDescription  string
+	helpText         string
+	helpMarkdown     string
+	resourceType     string
+	severity         string
+	url              string
+	resultIndex      int
+	artifactLocation string
+	message          string
+}
+
+func (sw *SarifWriter) addSarifRule(data *sarifData) {
+	r := sw.run.AddRule(data.vulnerabilityId).
+		WithName(toSarifRuleName(data.resourceType)).
+		WithDescription(data.vulnerabilityId).
+		WithFullDescription(&sarif.MultiformatMessageString{Text: &data.fullDescription}).
+		WithHelp(&sarif.MultiformatMessageString{
+			Text:     &data.helpText,
+			Markdown: &data.helpMarkdown,
+		}).
+		WithDefaultConfiguration(&sarif.ReportingConfiguration{
+			Level: toSarifErrorLevel(data.severity),
+		}).
+		WithProperties(sarif.Properties{
+			"tags": []string{
+				"vulnerability",
+				data.severity,
+			},
+			"precision": "very-high",
+		})
+	if data.url != "" {
+		r.WithHelpURI(data.url)
+	}
+}
+
+func (sw *SarifWriter) addSarifResult(data *sarifData) {
+	sw.addSarifRule(data)
+
+	location := sarif.NewPhysicalLocation().
+		WithArtifactLocation(sarif.NewSimpleArtifactLocation(data.artifactLocation).WithUriBaseId("ROOTPATH")).
+		WithRegion(sarif.NewRegion().WithStartLine(1))
+	result := sarif.NewRuleResult(data.vulnerabilityId).
+		WithRuleIndex(data.resultIndex).
+		WithMessage(sarif.NewTextMessage(data.message)).
+		WithLevel(toSarifErrorLevel(data.severity)).
+		WithLocations([]*sarif.Location{sarif.NewLocation().WithPhysicalLocation(location)})
+	sw.run.AddResult(result)
+}
+
+func getRuleIndex(id string, indexes map[string]int) int {
+	if i, ok := indexes[id]; ok {
+		return i
+	} else {
+		l := len(indexes)
+		indexes[id] = l
+		return l
+	}
+}
+
+func (sw SarifWriter) Write(report Report) error {
+	sarifReport, err := sarif.New(sarif.Version210)
+	if err != nil {
+		return err
+	}
+	sw.run = sarif.NewRunWithInformationURI("Trivy", "https://github.com/aquasecurity/trivy")
+	sw.run.Tool.Driver.WithVersion(sw.Version)
+	sw.run.Tool.Driver.WithFullName("Trivy Vulnerability Scanner")
+
+	ruleIndexes := map[string]int{}
+	for _, res := range report.Results {
+		for _, vuln := range res.Vulnerabilities {
+			fullDescription := vuln.Description
+			if fullDescription == "" {
+				fullDescription = vuln.Title
+			}
+
+			sw.addSarifResult(&sarifData{
+				vulnerabilityId:  vuln.VulnerabilityID,
+				severity:         vuln.Severity,
+				url:              vuln.PrimaryURL,
+				resourceType:     res.Type,
+				artifactLocation: toPathUri(res.Target),
```

Comment:
PkgPath can be used in case of individual packages.

Reply:
explicitly it didn't work, because `vuln.Path` is empty sometimes.
added a check:
```go
path := vuln.PkgPath
if path == "" {
	path = res.Target
}
```